### PR TITLE
gh-89051: Add ssl.OP_LEGACY_SERVER_CONNECT

### LINKED ISSUE #89051: ssl.OP_LEGACY_SERVER_CONNECT missing
BPO | [44888](https://bugs.python.org/issue44888)
--- | :---
Nosy | @tiran, @xtkoba
PRs | <li>python/cpython#27776</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/tiran'
closed_at = None
created_at = <Date 2021-08-11.11:35:40.075>
labels = ['expert-SSL']
title = 'ssl.OP_LEGACY_SERVER_CONNECT missing'
updated_at = <Date 2021-08-16.03:58:50.811>
user = 'https://github.com/xtkoba'
```

bugs.python.org fields:
```python
activity = <Date 2021-08-16.03:58:50.811>
actor = 'python-dev'
assignee = 'christian.heimes'
closed = False
closed_date = None
closer = None
components = ['SSL']
creation = <Date 2021-08-11.11:35:40.075>
creator = 'xtkoba'
dependencies = []
files = []
hgrepos = []
issue_num = 44888
keywords = ['patch']
message_count = 1.0
messages = ['399386']
nosy_count = 3.0
nosy_names = ['christian.heimes', 'python-dev', 'xtkoba']
pr_nums = ['27776']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue44888'
versions = []
```

</p></details>

### Linked PRs
* gh-93927
* gh-100359
* gh-100360

### PR DESCRIPTION
**Pull request by graingert**
Fixes #89051

* Issue: gh-89051

### CODE REVIEW

File: Doc/library/ssl.rst [hunk 1]
```diff
@@ -920,6 +920,14 @@ Constants
 
    .. versionadded:: 3.10
 
+
```
Comment (graingert) [inline Doc/library/ssl.rst]:
```suggestion
```

File: Lib/test/test_ssl.py [hunk 2]
```diff
@@ -4073,6 +4075,20 @@ def test_compression_disabled(self):
                                    sni_name=hostname)
         self.assertIs(stats['compression'], None)
 
+    def test_legacy_server_connect(self):
+        client_context, server_context, hostname = testing_context()
+        client_context.options |= ssl.OP_LEGACY_SERVER_CONNECT
+        server_params_test(client_context, server_context,
```
Comment (graingert) [inline Lib/test/test_ssl.py:3823]: Is there a mechanism for creating a bad TLS server that doesn't support secure renegotiation indication?

**Review by pquentin — APPROVED** [review]
Thanks! This will help with the transition to OpenSSL 3.0.

**Comment by graingert** [comment]
@tiran can I get a review on this please?

**Comment by tiran** [comment]
I guess you never found a why to create a server that does not support secure negotiations. Let's land this now anyway.

**Review by tiran — APPROVED** [review]

**Comment by graingert** [comment]
> I guess you never found a why to create a server that does not support secure negotiations. Let's land this now anyway.

I had a look into doing it with tlslite-ng or using openssl's or even just hand writing TLS down a plain socket.socket but it would add hundreds or thousands of lines to the PR

I did raise a ticket with badssl https://github.com/chromium/badssl.com/issues/507 if they implement it I'd be interested in integrating badssl.test as a docker container in the CPython CI.

**Comment by smontanaro** [comment]
(I'm working my way through some PRs which have been approved and are labeled "awaiting merge", hence my seemingly bolt from the blue comment. Why? [Read here](https://discuss.python.org/t/prs-awaiting-merge/21313).)

This has been idle since early August. @graingert Is there reason to assume a buggy TLS implementation is in the works against which this can be tested? If not, perhaps it's time to merge and close?

**Comment by graingert** [comment]
Yeah I think this just needs a merge

**Comment by dmpe** [comment]
Hi, any indication on when this will be merged & released ?

**Comment by hugovk** [comment]
Docs conflict resolved.

**Review by hugovk — APPROVED** [review]
Let's go for:

* `.. versionadded:: 3.10.10` in 3.10
* `.. versionadded:: 3.11.2` in 3.11
* `.. versionadded:: 3.12` in 3.12/main

**Comment by hugovk** [comment]
@dmpe Merged now, will be in [3.12.0 alpha 4, scheduled for 2023-01-09](https://peps.python.org/pep-0693/) and if the backports land soon, will be in [3.10.10](https://peps.python.org/pep-0619/) and [3.11.2](https://peps.python.org/pep-0664/), both scheduled for 2023-02-06.

**Comment by hugovk** [comment]
> @dmpe Merged now, will be in [3.12.0 alpha 4, scheduled for 2023-01-09](https://peps.python.org/pep-0693/) and if the backports land soon, will be in [3.10.10](https://peps.python.org/pep-0619/) and [3.11.2](https://peps.python.org/pep-0664/), both scheduled for 2023-02-06.

I closed the backport PRs, as we don't normally backport new features to bugfix branches.

https://devguide.python.org/versions/

This will be in the next 3.12.0 alpha 4, due on 2023-01-09.

**Pull request merged**